hcl2template/parser.go

```diff
@@ -293,112 +272,311 @@ func filterVarsFromLogs(inputOrLocal Variables) {
-			}
-			ref := datasource.Ref()
-			if existing, found := cfg.Datasources[ref]; found {
-				diags = append(diags, &hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  "Duplicate " + dataSourceLabel + " block",
-					Detail: fmt.Sprintf("This "+dataSourceLabel+" block has the "+
-						"same data type and name as a previous block declared "+
-						"at %s. Each "+dataSourceLabel+" must have a unique name per builder type.",
-						existing.block.DefRange.Ptr()),
-					Subject: datasource.block.DefRange.Ptr(),
-				})
-				continue
-			}
-			if cfg.Datasources == nil {
-				cfg.Datasources = Datasources{}
-			}
-			cfg.Datasources[ref] = *datasource
-		}
+	return diags
+}
+
+// decodeBuildBlock shallowly decodes a build block from the config.
+//
+// The final decoding step (which requires an up-to-date context) will be done
+// when we need it.
+func (cfg *PackerConfig) decodeBuildBlock(block *hcl.Block) hcl.Diagnostics {
+	build := &BuildBlock{
+		block: block,
+	}
+
+	cfg.Builds = append(cfg.Builds, build)
+
+	return nil
+}
+
+func (cfg *PackerConfig) decodeSourceBlock(block *hcl.Block) hcl.Diagnostics {
+	source, diags := cfg.decodeSource(block)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	if cfg.Sources == nil {
+		cfg.Sources = map[SourceRef]SourceBlock{}
+	}
+
+	ref := source.Ref()
+	if existing, found := cfg.Sources[ref]; found {
+		return append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Duplicate " + sourceLabel + " block",
+			Detail: fmt.Sprintf("This "+sourceLabel+" block has the "+
+				"same builder type and name as a previous block declared "+
+				"at %s. Each "+sourceLabel+" must have a unique name per builder type.",
+				existing.block.DefRange.Ptr()),
+			Subject: source.block.DefRange.Ptr(),
+		})
+	}
+
+	cfg.Sources[ref] = source
+
+	return diags
+}
+
+// decodeBuildSource reads a used source block from a build:
+//
+//	build {
+//	  source "type.example" {
+//	    name = "local_name"
+//	  }
+//	}
+func (cfg *PackerConfig) decodeBuildSource(block *hcl.Block) (SourceUseBlock, hcl.Diagnostics) {
+	ref := sourceRefFromString(block.Labels[0])
+	out := SourceUseBlock{SourceRef: ref}
+	var b struct {
+		Name string   `hcl:"name,optional"`
+		Rest hcl.Body `hcl:",remain"`
+	}
+	diags := gohcl.DecodeBody(block.Body, nil, &b)
+	if diags.HasErrors() {
+		return out, diags
+	}
+	out.LocalName = b.Name
+	out.Body = b.Rest
+	return out, nil
+}
+
+func (source *SourceBlock) finalizeDecodeSource(cfg *PackerConfig) hcl.Diagnostics {
+	if source.Ready {
+		return nil
+	}
+
+	source.Ready = true
+	dyn := dynblock.Expand(source.block.Body, cfg.EvalContext(DatasourceContext, nil))
+	// Expand without a base schema since nothing is known in advance for a
+	// source, but we still want to expand dynamic blocks if any
+	_, rem, diags := dyn.PartialContent(&hcl.BodySchema{})
+
+	// Only try to expand once, regardless of whether the source succeeded
+	// to expand dynamic data or not.
+	source.Ready = true
```

Comment:
Source is set to true on line 549. Should this be removed or the one above?

Reply:
Yes, I noticed this duplicate while working on the schedulers, I'll fix this one, thanks for calling it out